### DISCUSSION TITLE
[Support]: can Frigate (with its custom compiled ffmpeg) dewarp a "fisheye" lens' image?

### DISCUSSION THREAD

**Discussion by ElectronicBattle** [blakeblackshear/frigate#13090] [General Support] [answered]
### Describe the problem you are having

The native H.264 1536x1536 video stream from my My EZViz doorbell  looks like this:

![image](https://github.com/user-attachments/assets/4b33c902-a0c3-41e0-9b4b-e74ee5300cc1)

The Android EVZviz app processes it so that it looks like this:

![image](https://github.com/user-attachments/assets/15cc8a10-e080-479f-a5e2-d0048acec159)

**1.** Is Frigate able to  _directly_ achieve this sort of  dewarping? (I think not).
 
**2.** If not, is it something which would be done in conjunction with go2rtc to pipe a stream into a separate instance of ffmpeg perhaps, or vice versa? I don't understand (as usual) but I presume something like this:

`[doorbell]`**--->**`[Frigate+go2rtc]`**--->**`[external ffmpeg dewarper]`**--->**`[Frigate for onward processing]`

might be the way to approach the problem.

I tried it a couple of years ago in ffmpeg and ended up with every mapping known to man except the one I wanted. 

### Version

any

### Frigate config file

```yaml
n/a
```

### Relevant log output

```shell
n/a
```

### FFprobe output from your camera

```shell
ffprobe 'rtsp://admin:password@192.168.1.192:554/Streaming/Channels/101'

ffprobe version 4.4.5-0ubuntu1~22.04.sav0
Input #0, rtsp, from 'rtsp://admin:password@192.168.1.192:554/Streaming/Channels/101':
  Metadata:
    title           : Media Presentation
  Duration: N/A, start: 0.000000, bitrate: N/A
  Stream #0:0: Video: h264 (Main), yuv420p(progressive), 1536x1536, 15 fps, 14.99 tbr, 90k tbn, 30 tbc
  Stream #0:1: Audio: aac (LC), 16000 Hz, mono, fltp
```

### Frigate stats

```json
n/a
```

### Operating system

Debian

### Install method

Docker Compose

### Object Detector

Coral

### Network connection

Wired

### Camera make and model

EZViz DB1C https://www.ezviz.com/product/db1c/27224

**Comment by NickM-27** [comment, marked as answer]
This can be done in Frigate via ffmpeg https://github.com/blakeblackshear/frigate/issues/9614#issuecomment-1924927994

**Reply by ElectronicBattle** [reply]
Astonishing! Is there any thing you _can't_ get it to do?

Thanks - more stuff to play with. I am busy integrating the remaining four out of my eight cameras, plus this one if I can dewarp successfully (_"Captain: Klingons off the starboard bow"_)

**Reply by mbooij** [reply]
Dit u get this to work? If so, could u please share your frigate config?

**Reply by ElectronicBattle** [reply]
This thread sorted the problems out for me:

https://github.com/blakeblackshear/frigate/discussions/17013

and my own config has this section in it for `go2rtc`:

```
doorbell:
      - ffmpeg:rtsp://admin:password@192.168.1.192:554/Streaming/Channels/101#video=h264#audio=copy#raw=-vf
        "lenscorrection=k1=-0.450:k2=0.1"                                                                                                   #new from hawkeye
      - ffmpeg:doorbell#audio=opus
  ```
You will see from the screenshots in that thread that it works very well for me. 

Good luck!

**Reply by mbooij** [reply]
Thank you. I will try to replicate this for my EZViz DB1C. Currently i have a low res 1:1 stream up that allready eats most of my CPU/hardware acceleration does not seem to be utilized for the lenscorrection.

**Reply by ElectronicBattle** [reply]
How did you get a low res stream? My two streams are both high resolution as you will see from the screenshots below.

My cameras portion for my doorbell says

```
###############doorbell#############################
  doorbell:
    genai:
      enabled: false # <- enable GenAI for this camera
    enabled: true
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/doorbell 
          roles:
            - detect
        - path: rtsp://127.0.0.1:8554/doorbell 
          roles:
            - record
```

and I see that the DB1C stream is high resolution (1536x1536) and uses a couple of GB/hour bandwidth compared with my ordinary 1920x1080p 2 megapixel cameras which are a tenth of that rate.

![image](https://github.com/user-attachments/assets/59e2e933-4f08-485a-b541-d4d162c127f1)

![image](https://github.com/user-attachments/assets/a59e5749-9ed2-4f99-96de-ff11c527c029)

![image](https://github.com/user-attachments/assets/3b629bc8-9cd9-4e25-b15a-45353c43ffc4)

**Reply by mbooij** [reply]
My current cobbled together "working" config is:
`  ezviz_fisheye:
    ffmpeg:
      inputs:
        - path: rtsp://admin:password@192.168.x.x:554/h264_stream
          roles:
            - detect
            - record
      output_args:
        record: -vf lenscorrection=cx=0.5:cy=0.5:k1=-0.290:k2=-0.012 -f segment -segment_time 10 -segment_format mp4 -reset_timestamps 1 -strftime 1 -ar 44100 -c:a aac
        detect: -vf lenscorrection=cx=0.5:cy=0.5:k1=-0.290:k2=-0.012 -f rawvideo -pix_fmt yuv420p
    detect:
      width: 1536
      height: 1536
    objects:
      track:
        - person`
Dewarping works "ok" but the resolution in Frigate is very bad compared to an rtsp source-stream via VLC. The aspect ratio is also very off compared to your screenshots. Hoping to get an end results simular to non fisheye camera's. This config also taxes my NVR cpu and is (probably as a result of high cpu load) unstable.  Will try to replicate your config with adding a go2rtc section. This is new for me since no other camera's have this fisheye effect. Will keep u updated.
![image](https://github.com/user-attachments/assets/16670f28-8d23-4a13-912c-ab0b5fa8b364)

**Reply by ElectronicBattle** [reply]
Good luck.  It takes me *ages* to get this sort of thing working because I largely don't know what I am doing and I have very little skill in this sort of area. Having said that, if I can help I will certainly try!

**Reply by mbooij** [reply]
I got it working "better" with a seperate go2rtc section for the lenscorrection. Is there a perticular reason for separating the streams for detecting and recording? This is my current config: 
`go2rtc:
  log:
    exec: trace
  streams:
       videodoorbell:
        - ffmpeg:rtsp://admin:password@192.168.x.x:554/Streaming/Channels/101#video=h264#audio=copy#raw=-vf "lenscorrection=k1=-0.450:k2=0.1"
        - ffmpeg:videodoorbell#audio=opus
  webrtc:
    candidates:
      - 192.168.0.60:8555
      - stun:8555

cameras:
  videodoorbell:
    enabled: true
    ffmpeg:
      inputs:
        - path: rtsp://127.0.0.1:8554/videodoorbell
          roles:
            - detect
            - record
    detect:
      width: 1536
      height: 1536
      fps: 5
    objects:
      track:
        - person`
This fixes the low resolution issue from the former config. The videofeed i still is very unstable for me dropping out sometimes. Perhaps this is due to the 2.4Ghz WiFi connetion. Will try to run the doorbell on 5Ghz soon.

**Reply by ElectronicBattle** [reply]
`- path: rtsp://127.0.0.1:8554/videodoorbel`  this should say

`- path: rtsp://127.0.0.1:8554/videodoorbell`

so that the name matches the description in the go2rtc section.  (Perhaps it is just a typo error here rather than in the actual config).

I think the dewarping has to happen in the go2rtc section because, effectively, go2rtc receives the video stream and then "retransmits" the video traffic having applied the dewarping filter to it.

**Reply by mbooij** [reply]
This was an translation / modification error on my side. I am using localised names for my camera's and made an error. I corrected this in the code above. It was not an issue in my local code. I also added the webrtc section for smooth usage in homeassistant. This doorbell now works great with this configuration. Thanks u for your support.

**Reply by ElectronicBattle** [reply]
(_and a minor update in case it is useful for you: I found today that `k1=-0.6:k2=0.2` improved the linearity of the dewarp so that the resultant picture is even better)_.
